### PR TITLE
viewer : restrict metadata searches to a specified extent, optionnally set that extent to the current map extent

### PR DESCRIPTION
**Pull request by fphg**
No GUI for now, this is admin's choice

### CODE REVIEW

File: mapfishapp/src/main/webapp/app/js/GEOR_cswquerier.js [hunk 1]
```diff
@@ -826,32 +849,47 @@ Ext.app.FreetextField = Ext.extend(Ext.form.TwinTriggerField, {
...
-                ]
-            })
+        // combine all filters alltogether
+        finalFilters.push(byTypes);
+        
+        // spatial filter
+        if (GEOR.config.CSW_FILTER_SPATIAL.length!==4) {
+            finalFilters.push(new OpenLayers.Filter.Spatial({
+                    type: OpenLayers.Filter.Spatial.BBOX,
+                    value: this.map.getExtent().clone()
```
Comment (fvanderbiest) [inline mapfishapp/src/main/webapp/app/js/GEOR_cswquerier.js]: no need to clone extent here IMO

File: mapfishapp/src/main/webapp/app/js/GEOR_config.js [hunk 2]
```diff
@@ -229,6 +229,14 @@ GEOR.config = (function() {
         ]),
 
         /**
+         * Constant: CSW_FILTER_SPATIAL
+         * An optional extent in latlon to restrict metadata search on a specific extent (latlon)
+         * Defaults to [-180,-90,180,90] to cover the world.
+         * If the parameter is not an array of 4, the filter uses the current map extent.
+         */
+        CSW_FILTER_SPATIAL: getCustomParameter("CSW_FILTER_SPATIAL", [-180,-90,180,90]),
```
Comment (fvanderbiest) [inline mapfishapp/src/main/webapp/app/js/GEOR_config.js]: I would rather make this null by default.

File: mapfishapp/src/main/webapp/app/js/GEOR_cswquerier.js [hunk 3]
```diff
@@ -826,32 +849,47 @@ Ext.app.FreetextField = Ext.extend(Ext.form.TwinTriggerField, {
...
-                            })
-                        ]
-                    })
-                ]
-            })
+        // combine all filters alltogether
+        finalFilters.push(byTypes);
+        
+        // spatial filter
+        if (GEOR.config.CSW_FILTER_SPATIAL.length!==4) {
```
Comment (fvanderbiest) [inline mapfishapp/src/main/webapp/app/js/GEOR_cswquerier.js]: if not null, then ... (we assume that the admin knows what he does)

**Comment by fvanderbiest** [comment]
Some minor comments to be addressed.

Finally, I think we need to inform the user somehow that his search is/might be restricted on the current bbox.

File: mapfishapp/src/main/webapp/app/js/GEOR_cswquerier.js (hunk 1 above)
Comment (fphg) [inline mapfishapp/src/main/webapp/app/js/GEOR_cswquerier.js]: dunno if we'll expand or alter this bbox later

File: mapfishapp/src/main/webapp/app/js/GEOR_config.js (hunk 2 above)
Comment (fphg) [inline mapfishapp/src/main/webapp/app/js/GEOR_config.js]: right

**Comment by fphg** [comment]
fvanderbiest : now there's a warning in the search result report and a dedicated checkbox to enable/disable the auto bbox feature on the fly. this comes in handy at querying large external catalogs. Repeated queries trigger annoyingly the search tooltip, so PR makes it pop only once.

File: mapfishapp/src/main/webapp/app/js/GEOR_cswquerier.js [hunk 4]
```diff
@@ -462,6 +485,57 @@ GEOR.cswquerier = (function() {
...
+        /**
+         * APIMethod: getSpatialFilter
+         * Return the search extent to build a csw search filter
+         *
+         * Returns:
+         * {OpenLayers.Filter.Spatial} bbox filter
+         */
+        getSpatialFilter: function() {
+            var f;
+            if (isLimitExtent) {
```
Comment (fvanderbiest) [inline mapfishapp/src/main/webapp/app/js/GEOR_cswquerier.js]: I would factorize this contruct, in order to define the filter only once.
Its value should be set with eg:

```js
var v = isLimitExtent ? 
    map.getExtent().clone().transform(
        map.getProjectionObject(), 
        new OpenLayers.Projection("EPSG:4326")
    ) :
    new OpenLayers.Bounds(
         GEOR.config.CSW_FILTER_SPATIAL
    )
```

File: mapfishapp/src/main/webapp/app/js/GEOR_cswquerier.js [hunk 5]
```diff
@@ -193,6 +194,24 @@ GEOR.cswquerier = (function() {
...
+     * Property: isLimitExtent
+     * {Boolean} limit metadata results to map extent
+     */
+    var isLimitExtent = false;
+    
+    /**
+     * Property: isTooltip
+     * {Boolean} display search tooltip only once
+     */
+    var isTooltip = true;
```
Comment (fvanderbiest) [inline mapfishapp/src/main/webapp/app/js/GEOR_cswquerier.js]: rather ```tipDisplayed``` with a default value to false ?

File: mapfishapp/src/main/webapp/app/js/GEOR_cswquerier.js (hunk 4 above)
Comment (fphg) [inline mapfishapp/src/main/webapp/app/js/GEOR_cswquerier.js]: done, thanks for the advice

File: mapfishapp/src/main/webapp/app/js/GEOR_cswquerier.js (hunk 5 above)
Comment (fphg) [inline mapfishapp/src/main/webapp/app/js/GEOR_cswquerier.js]: thanks, renamed ```isTipDisplayed```. don't we want the tip to be displayed on 1st search ?

File: mapfishapp/src/main/webapp/app/js/GEOR_cswquerier.js [hunk 6]
```diff
@@ -462,6 +485,55 @@ GEOR.cswquerier = (function() {
...
+        /**
+         * APIMethod: init
+         * Initialize this module 
+         *
+         * Parameters:
+         * m - {OpenLayers.Map} The map instance.
+         */
+        init: function(m) { 
+            map = m;
+            isLimitExtent = (GEOR.config.CSW_FILTER_SPATIAL===null);
```
Comment (fvanderbiest) [inline mapfishapp/src/main/webapp/app/js/GEOR_cswquerier.js]: I would rename it to ```dynamicSpatialFilter``` as well

File: mapfishapp/src/main/webapp/app/js/GEOR_cswquerier.js [hunk 7]
```diff
@@ -193,6 +194,24 @@ GEOR.cswquerier = (function() {
...
+     * Property: map
+     * {OpenLayers.Map} The map instance.
+     */
+    var map = null;
+    
+    /**
+     * Property: isLimitExtent
+     * {Boolean} limit metadata results to map extent
+     */
+    var isLimitExtent = false;
```
Comment (fvanderbiest) [inline mapfishapp/src/main/webapp/app/js/GEOR_cswquerier.js]: In fact, no need to initialize it to anything, since this is done in init()

**Comment by fvanderbiest** [comment]
Still some minor comments to be addressed.
Otherwise, looks good. Thanks for the contribution :-)

**Comment by fphg** [comment]
is it ok now ?

**Comment by fvanderbiest** [comment]
Yes, thanks again

**Comment by fvanderbiest** [comment]
I think this PR is incomplete.
When the user has checked the box, he expects the results to change on map extent changed => https://github.com/georchestra/georchestra/issues/1007

**Pull request merged**